File: src/MailSearch.Core/Organizer/Organizer.cs
using System.Collections.Generic;
using MailSearch.Core.Database;
using MailSearch.Core.Models;

namespace MailSearch.Core.Organizer;

/// <summary>
/// Provides tagging and listing operations for emails.
/// </summary>
public class Organizer(MailSearchRepository repo)
{
    /// <summary>Adds one or more tags to an email.</summary>
    public void Tag(int emailId, params string[] tagNames)
    {
        foreach (var name in tagNames)
            repo.AddTagToEmail(emailId, name.Trim());
    }

    /// <summary>Removes a tag from an email.</summary>
    public void Untag(int emailId, string tagName)
    {
        repo.RemoveTagFromEmail(emailId, tagName.Trim());
    }

    /// <summary>Returns all tags on a specific email.</summary>
    public List<Tag> GetTagsForEmail(int emailId) =>
        repo.GetTagsForEmail(emailId);

    /// <summary>Returns all tags in the database.</summary>
    public List<Tag> ListTags() =>
        repo.ListTags();

    /// <summary>Returns emails that have a given tag.</summary>
    public List<Email> ListEmailsByTag(string tagName, int limit = 50, int offset = 0) =>
        repo.ListEmails(limit, offset, tag: tagName);
}

File: src/organizer/organizer.ts
import { MailSearchRepository } from '../database/repository';
import { Tag } from '../models/tag';
import { Email } from '../models/email';

/**
 * Organizer provides tagging and listing operations for emails.
 */
export class Organizer {
  constructor(private repo: MailSearchRepository) {}

  /** Add one or more tags to an email. */
  tag(emailId: number, ...tagNames: string[]): void {
    for (const name of tagNames) {
      this.repo.addTagToEmail(emailId, name.trim());
    }
  }

  /** Remove a tag from an email. */
  untag(emailId: number, tagName: string): void {
    this.repo.removeTagFromEmail(emailId, tagName.trim());
  }

  /** List all tags on a specific email. */
  getTagsForEmail(emailId: number): Tag[] {
    return this.repo.getTagsForEmail(emailId);
  }

  /** List all tags in the database. */
  listTags(): Tag[] {
    return this.repo.listTags();
  }

  /** List emails with a specific tag. */
  listEmailsByTag(tagName: string, limit = 50, offset = 0): Email[] {
    return this.repo.listEmails({ tag: tagName, limit, offset });
  }
}
